src/config.rs

```diff
@@ -195,14 +220,94 @@ impl ConfigManager {
         Ok(())
     }
 
-    pub fn get_or_create_config(&self) -> Result<Config> {
-        let config_file_path = self.config_file_path();
-        if !config_file_path.is_file() {
-            Config::create_default_config_to_path(&config_file_path)?;
+    pub fn get_config(&self) -> Option<&Config> {
+        if let Some(config) = &self.config {
+            Some(&config)
+        } else {
+            None
+        }
+    }
+
+    pub fn parse_authorized_cert(&self) -> Result<AuthorizedCerts> {
+        let authorized_certs_path = self.authorized_certs_path();
+        if !authorized_certs_path.is_file() {
+            fs::File::create(authorized_certs_path)?;
+            return Ok(AuthorizedCerts::new());
+        }
+        AuthorizedCerts::from_file(authorized_certs_path)
+    }
+
+    pub fn add_authorized_cert(&self, cert_der: &dyn AsRef<[u8]>) -> Result<()> {
+        let cert_der = cert_der.as_ref();
+        let digest = ring::digest::digest(&ring::digest::SHA256, cert_der);
+        let digest_hex_formatted =
+            digest.as_ref().iter()
+            .map(|x| format!("{:02X}", x))
+            .collect::<Vec<String>>()
+            .join(":");
+        let authorized_certs = self.parse_authorized_cert()?;
+        if authorized_certs.certs.get(&digest_hex_formatted).is_some() {
+            bail!("client certificate already exists")
         }
-        let config_content = fs::read_to_string(&config_file_path)?;
-        let mut ini = Ini::new_cs();
-        ini.read(config_content).map_err(Error::msg)?;
-        Config::from_ini(&ini)
+
+        let digest_hex_formatted = format!("SHA256:{}", digest_hex_formatted);
+        let base64_str = base64::encode(cert_der);
+        let mut f = fs::OpenOptions::new()
+            .write(true)
+            .append(true)
+            .open(self.authorized_certs_path())?;
+        let s = format!("{} {}\n", digest_hex_formatted, base64_str);
+        f.write(&s.into_bytes())?;
+        Ok(())
+    }
+}
+
+// data structure that holds `authorized_certs` file content
+pub struct AuthorizedCerts {
+    // Fingerprint -> Certificate map
+    certs: HashMap<String, Vec<u8>>
+}
+
+impl AuthorizedCerts {
+    fn new() -> AuthorizedCerts {
+        AuthorizedCerts {
+            certs: HashMap::new(),
+        }
+    }
+
+    fn from_file(path: PathBuf) -> Result<AuthorizedCerts> {
```

Comment:
```suggestion
    fn from_path(path: &Path) -> Result<AuthorizedCerts> {
```

from_path가 더 맞는 표현인 것 같아요. 그리고 이건 체크해보진 않았는데 &Path만 받아도 된다면 owned 타입인 PathBuf보다 &Path가 나을거에요

Reply:
```suggestion
    fn from_file(path: PathBuf) -> Result<Self> {
```

아 글고 중요한건 아닌데 타입을 Self로 적으면 보기에도 편하고 나중에 혹시 리팩토링할 때도 편해진답니다

Reply:
8b076fd